chiron/multistate.py

```diff
@@ -0,0 +1,968 @@
+
+        Returns
+        -------
+        np.ndarray
+            An array of updated thermodynamic state indices for each replica.
+        """
+
+        log.debug("Mixing replicas (does nothing for MultiStateSampler)...")
+
+        # Reset storage to keep track of swap attempts this iteration.
+        self._n_accepted_matrix[:, :] = 0
+        self._n_proposed_matrix[:, :] = 0
+
+        # Perform replica mixing (swap proposals and acceptances)
+        # The actual swapping logic would depend on subclass implementations
+        # Here, we assume a placeholder implementation
+        new_replica_states = self._perform_swap_proposals()
+
+        # Calculate swap acceptance statistics
+        n_swaps_proposed = self._n_proposed_matrix.sum()
+        n_swaps_accepted = self._n_accepted_matrix.sum()
+        swap_fraction_accepted = 0.0
+        if n_swaps_proposed > 0:
+            swap_fraction_accepted = n_swaps_accepted / n_swaps_proposed
+        log.debug(
+            f"Accepted {n_swaps_accepted}/{n_swaps_proposed} attempted swaps ({swap_fraction_accepted * 100.0:.1f}%)"
+        )
+        return new_replica_states
+
+    @with_timer("Propagating all replicas")
+    def _propagate_replicas(self) -> None:
+        """
+        Propagate all replicas through their respective MCMC moves.
+
+        This method iterates over all replicas and applies the corresponding MCMC move
+        to each one, based on its current thermodynamic state.
+        """
+
+        log.debug("Propagating all replicas...")
+
+        for replica_id in range(self.n_replicas):
+            self._propagate_replica(replica_id)
+
+    def _neighborhood(self, state_index: int) -> List[int]:
+        """
+        Compute the indices of neighboring states for a given state.
+
+        This method determines the neighborhood of states around a given state index,
+        considering the 'locality' parameter. If 'locality' is None, the neighborhood
+        includes all states; otherwise, it includes states within the 'locality' range.
+
+        Parameters
+        ----------
+        state_index : int
+            The index of the state for which the neighborhood is to be calculated.
+
+        Returns
+        -------
+        List[int]
+            A list of state indices that are considered neighbors of the given state.
+        """
+        if self.locality is None:
+            # Global neighborhood
+            return list(range(self.n_states))
+        else:
+            # Local neighborhood specified by 'locality'
+            lower_bound = max(0, state_index - self.locality)
+            upper_bound = min(self.n_states, state_index + self.locality + 1)
+            return list(
+                range(
+                    lower_bound,
+                    upper_bound,
+                )
+            )
+
+    @with_timer("Computing energy matrix")
+    def _compute_energies(self) -> None:
```

Comment:
We might want to be careful with this method since we will want to `jit` the inner energy matrix computation loop, and potentially introduce `vmap/pmap` parallelization.

We also have to figure out if we will ever want to `jit` the whole call to `run`, or if we're fine with just `jit`ting the individual time-consuming substeps (replica swapping, replica propagation, energy matrix computation),

Reply:
This has been addressed in PR #21